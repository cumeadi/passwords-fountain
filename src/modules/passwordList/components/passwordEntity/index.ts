export { PasswordEntity } from './passwordEntity.component';
