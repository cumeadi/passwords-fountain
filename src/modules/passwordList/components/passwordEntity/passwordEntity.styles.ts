import { h } from 'preact';
import { styled, setPragma } from 'goober';
import { theme, typography } from '@/common/theme';
setPragma(h);

export const Row = styled('div')`
    margin-bottom: ${theme.spacing.m18};
`;

export const Wrapper = styled('section')`
    background: ${theme.colors.pastelGreen};
    padding: ${theme.spacing.m18};

    ${Row}:last-of-type {
        margin-bottom: 0;
    }
`;
export const Label = styled('span')`
    font-family: ${theme.fontFamilies.bold};
    color: ${theme.colors.darkBlue};
    ${typography.text18}
`;

export const Value = styled('strong')`
    color: ${theme.colors.darkBlue};
    word-break: break-all;
    ${typography.text18}
`;
