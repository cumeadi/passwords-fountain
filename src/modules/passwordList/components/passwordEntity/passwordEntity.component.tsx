import { h, Fragment } from 'preact';
import { TypedComponent } from '@/common/typings/prop-types';
import PropTypes from 'prop-types';
import {
    Wrapper,
    GridWrapper,
    Row,
    Label,
    Value,
    DataWrapper,
    ControlsWrapper,
    FormControlWrapper,
} from './passwordEntity.styles';
import { Text } from '@/modules/localisation/components/text';
import {
    placeholderEntityValue,
    PasswordEntityVulnerablePayload,
} from '@/modules/passwordList/passwordList.constants';
import { useRef } from 'preact/hooks';
import { PasswordEntityRaw } from '@/modules/database/database.service';
import { IconButton } from '@/common/components/iconButton';
import { renderIfTrue } from '@/common/utils/rendering';
import { Prompt } from '@/modules/overlay/components/prompt';
import { useInputFormControl } from '@/common/utils/form';
import { masterKey } from '@/common/utils/formValidators';
import { FormControl } from '@/common/components/formControl';
import { TextInput } from '@/common/components/textInput';
import { Button } from '@/common/components/button';
import { useAction, useSelector } from '@/store';
import { passwordListActions } from '@/modules/passwordList/passwordList.actions';
import { selectSelectedAndDecryptedEntityByRefId } from '@/modules/passwordList/passwordList.selectors';

const formValidation = { masterKey } as const;

export const PasswordEntity: TypedComponent<Props> = ({
    data,
    isSelected,
    onClick,
    promptVisibility,
    setPromptVisibility,
}: Props) => {
    const formRef = useRef(undefined as any);
    const setSelectedAndDecryptedEntity = useAction(
        passwordListActions.setSelectedAndDecryptedEntity
    );
    const resetSelectedAndDecryptedEntity = useAction(
        passwordListActions.resetSelectedAndDecryptedEntity
    );
    const selectedAndDecryptedEntity = useSelector(
        selectSelectedAndDecryptedEntityByRefId(data.ref.id)
    );
    const passwordVisibility =
        Object.keys(selectedAndDecryptedEntity).length !== 0;

    const [masterKeyInputState, masterKeyInputProps] = useInputFormControl(
        formRef,
        formValidation,
        'masterKey'
    );

    const handleClick = (): void => {
        onClick(isSelected ? null : data);
        resetSelectedAndDecryptedEntity();
        setPromptVisibility(false);
        masterKeyInputState.setValue('');
    };

    const handlePromptConfirm = async (): Promise<void> => {
        const { decrypt } = await import('@/modules/cipher/cipher.service');
        const { login, password } = decrypt(
            data.data.value,
            masterKeyInputState.value,
            true
        ) as PasswordEntityVulnerablePayload;

        setSelectedAndDecryptedEntity({
            refId: data.ref.id,
            label: data.data.label,
            login,
            password,
        });
        setPromptVisibility(false);
        masterKeyInputState.setValue('');
    };

    const handlePromptCancel = (): void => {
        setPromptVisibility(false);
        masterKeyInputState.setValue('');
    };

    const handleEyeClick = (e: Event): void => {
        e.stopPropagation();
        if (promptVisibility) {
            return;
        }

        setPromptVisibility(true);
    };

    const handleFilledEyeClick = (e: Event): void => {
        e.stopPropagation();
        resetSelectedAndDecryptedEntity();
    };

    const renderDecryptionPrompt = renderIfTrue(() => (
        <Prompt
            renderContent={() => (
                <form ref={formRef}>
                    <FormControlWrapper>
                        <FormControl
                            hasError={masterKeyInputProps.hasError}
                            renderLabel={() => (
                                <Text>optionsPanel.enterMasterKey</Text>
                            )}
                            renderError={() => (
                                <Text>{masterKeyInputState.errors}</Text>
                            )}
                            renderInput={() => (
                                <TextInput
                                    type="password"
                                    placeholder="e.g. MyStrongPassword1234"
                                    {...masterKeyInputProps}
                                />
                            )}
                        />
                    </FormControlWrapper>
                </form>
            )}
            renderControls={() => (
                <Fragment>
                    <Button onClick={handlePromptCancel}>
                        <Text>optionsPanel.cancel</Text>
                    </Button>
                    <Button
                        onClick={handlePromptConfirm}
                        disabled={masterKeyInputProps.hasError}
                    >
                        <Text>optionsPanel.confirm</Text>
                    </Button>
                </Fragment>
            )}
        />
    ));

    const renderEyeIcon = () => {
        if (passwordVisibility) {
            return (
                <IconButton
                    iconName="eyeFilled"
                    onClick={handleFilledEyeClick}
                />
            );
        }
        return <IconButton iconName="eye" onClick={handleEyeClick} />;
    };

    const renderControls = renderIfTrue(() => {
        return (
            <Fragment>
                <IconButton
                    iconName="bin"
                    onClick={(e: any): void => {
                        e.stopPropagation();
                    }}
                />
                {renderEyeIcon()}
            </Fragment>
        );
    });

    return (
        <Wrapper onClick={handleClick} isSelected={isSelected}>
            <GridWrapper>
                <DataWrapper>
                    <Row>
                        <Label>
                            <Text>passwordEntity.label</Text>
                        </Label>{' '}
                        - <Value>{data.data.label}</Value>
                    </Row>
                    <Row>
                        <Label>
                            <Text>passwordEntity.login</Text>
                        </Label>{' '}
                        -{' '}
                        <Value>
                            {selectedAndDecryptedEntity.login ??
                                placeholderEntityValue}
                        </Value>
                    </Row>
                    <Row>
                        <Label>
                            <Text>passwordEntity.password</Text>
                        </Label>{' '}
                        -{' '}
                        <Value>
                            {selectedAndDecryptedEntity.password ??
                                placeholderEntityValue}
                        </Value>
                    </Row>
                </DataWrapper>
                <ControlsWrapper>{renderControls(isSelected)}</ControlsWrapper>
            </GridWrapper>
            {renderDecryptionPrompt(promptVisibility && isSelected)}
        </Wrapper>
    );
};

interface Props {
    isSelected: boolean;
    data: PasswordEntityRaw;
    onClick: Function;
    promptVisibility: boolean;
    setPromptVisibility: Function;
}

PasswordEntity.propTypes = {
    isSelected: PropTypes.bool.isRequired,
    data: PropTypes.any.isRequired,
    onClick: PropTypes.func.isRequired,
    promptVisibility: PropTypes.bool.isRequired,
    setPromptVisibility: PropTypes.func.isRequired,
};
