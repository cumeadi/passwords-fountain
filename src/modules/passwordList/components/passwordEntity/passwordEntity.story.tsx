import { h, VNode } from 'preact';
import { PasswordEntity } from './passwordEntity.component';
import { withA11y } from '@storybook/addon-a11y';

export default {
    title: 'PasswordEntity',
    decorators: [withA11y],
};

export const defaultView = (): VNode => (
    <PasswordEntity
        data={{
            ref: { id: 'x1', value: { id: 'x1 ' } },
            ts: Date.now(),
            data: { label: 'My EA Account', value: 'v1' },
        }}
        isSelected={false}
        onClick={() => {
            /* eslint-disable @typescript-eslint/no-empty-function */
        }}
        promptVisibility={false}
        setPromptVisibility={() => {
            /* eslint-disable @typescript-eslint/no-empty-function */
        }}
    />
);
