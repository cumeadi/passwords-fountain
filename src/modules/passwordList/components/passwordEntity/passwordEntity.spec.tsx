import '@testing-library/jest-dom/extend-expect';
import { h } from 'preact';
import { render, fireEvent } from '@testing-library/preact';
import { PasswordEntity } from './passwordEntity.component';
import { StoreProvider } from '@preact-hooks/unistore';
import createStore from 'unistore';

const stateMock = {
    passwordList: {
        selectedAndDecryptedEntity: {},
    },
};

const dataMock = {
    ref: { id: 'mock_ref_id', value: { id: 'mock_ref_id' } },
    ts: Date.now(),
    data: {
        label: 'My EA Account',
        value: 'mock_encrypted_login_and_password',
    },
};

describe('PasswordEntity', () => {
    afterEach(() => {
        jest.restoreAllMocks();
    });

    describe('when clicked', () => {
        it('calls onClick callback', () => {
            const onClickCb = jest.fn();

            const { queryAllByRole } = render(
                <StoreProvider value={createStore(stateMock)}>
                    <PasswordEntity
                        data={dataMock}
                        isSelected={false}
                        onClick={onClickCb}
                    />
                </StoreProvider>
            );

            fireEvent(
                queryAllByRole('button')[0],
                new MouseEvent('click', { detail: 1 })
            );
            expect(onClickCb).toHaveBeenCalledTimes(1);
        });
    });

    describe('when not selected', () => {
        it('renders', () => {
            const { asFragment } = render(
                <StoreProvider value={createStore(stateMock)}>
                    <PasswordEntity
                        data={dataMock}
                        isSelected={false}
                        onClick={() => {
                            /* eslint-disable @typescript-eslint/no-empty-function */
                        }}
                    />
                </StoreProvider>
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });

    describe('when selected', () => {
        it('renders with additional icon buttons', () => {
            const { asFragment } = render(
                <StoreProvider value={createStore(stateMock)}>
                    <PasswordEntity
                        data={dataMock}
                        isSelected
                        onClick={() => {
                            /* eslint-disable @typescript-eslint/no-empty-function */
                        }}
                    />
                </StoreProvider>
            );

            expect(asFragment()).toMatchSnapshot();
        });

        describe('when eye icon button has been clicked', () => {
            it('renders prompt', async () => {
                const onClickCb = jest.fn();

                const { queryAllByRole, asFragment } = render(
                    <StoreProvider value={createStore(stateMock)}>
                        <PasswordEntity
                            data={dataMock}
                            isSelected
                            onClick={onClickCb}
                        />
                    </StoreProvider>
                );

                fireEvent.click(queryAllByRole('button')[2]);

                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe('when bin icon button has been clicked', () => {
            it('renders prompt', async () => {
                const onClickCb = jest.fn();

                const { queryAllByRole, asFragment } = render(
                    <StoreProvider value={createStore(stateMock)}>
                        <PasswordEntity
                            data={dataMock}
                            isSelected
                            onClick={onClickCb}
                        />
                    </StoreProvider>
                );

                fireEvent.click(queryAllByRole('button')[1]);
                expect(asFragment()).toMatchSnapshot();
            });
        });
    });
});
