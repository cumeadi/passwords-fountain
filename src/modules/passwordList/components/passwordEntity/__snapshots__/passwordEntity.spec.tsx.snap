// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordEntity when not selected renders 1`] = `
<DocumentFragment>
  <button
    class="go113998"
  >
    <span
      class="go1470750214"
    >
      <div
        class="go4131202203"
      >
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Label
          </span>
           - 
          <strong
            class="go1914257566"
          >
            My EA Account
          </strong>
        </div>
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Login
          </span>
           - 
          <strong
            class="go1914257566"
          >
            XXX
          </strong>
        </div>
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Password
          </span>
           - 
          <strong
            class="go1914257566"
          >
            XXX
          </strong>
        </div>
      </div>
      <div
        class="go1569473757"
      />
    </span>
  </button>
</DocumentFragment>
`;

exports[`PasswordEntity when selected renders with additional icon buttons 1`] = `
<DocumentFragment>
  <button
    class="go1778638100"
    isselected="true"
  >
    <span
      class="go1470750214"
    >
      <div
        class="go4131202203"
      >
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Label
          </span>
           - 
          <strong
            class="go1914257566"
          >
            My EA Account
          </strong>
        </div>
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Login
          </span>
           - 
          <strong
            class="go1914257566"
          >
            XXX
          </strong>
        </div>
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Password
          </span>
           - 
          <strong
            class="go1914257566"
          >
            XXX
          </strong>
        </div>
      </div>
      <div
        class="go1569473757"
      />
    </span>
  </button>
  <button
    class="go3135434547 go3618827259"
  >
    <span
      class="go4062099715"
    >
      <img
        alt="icon"
        class="go115679"
        height="30"
        src="static-asset-stub"
        width="30"
      />
    </span>
  </button>
  <button
    class="go3135434547 go3618827259"
  >
    <span
      class="go4062099715"
    >
      <img
        alt="icon"
        class="go115679"
        height="30"
        src="static-asset-stub"
        width="30"
      />
    </span>
  </button>
</DocumentFragment>
`;

exports[`PasswordEntity when selected when bin icon button has been clicked renders prompt 1`] = `
<DocumentFragment>
  <button
    class="go1778638100"
    isselected="true"
  >
    <span
      class="go1470750214"
    >
      <div
        class="go4131202203"
      >
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Label
          </span>
           - 
          <strong
            class="go1914257566"
          >
            My EA Account
          </strong>
        </div>
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Login
          </span>
           - 
          <strong
            class="go1914257566"
          >
            XXX
          </strong>
        </div>
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Password
          </span>
           - 
          <strong
            class="go1914257566"
          >
            XXX
          </strong>
        </div>
      </div>
      <div
        class="go1569473757"
      />
    </span>
  </button>
  <button
    class="go3135434547 go3618827259"
  >
    <span
      class="go4062099715"
    >
      <img
        alt="icon"
        class="go115679"
        height="30"
        src="static-asset-stub"
        width="30"
      />
    </span>
  </button>
  <button
    class="go3135434547 go3618827259"
  >
    <span
      class="go4062099715"
    >
      <img
        alt="icon"
        class="go115679"
        height="30"
        src="static-asset-stub"
        width="30"
      />
    </span>
  </button>
  <section
    class="go939053875"
  >
    <div
      class="go579113919"
    >
      <form>
        <div
          class="go4083855640"
        >
          <section
            class="go3154290440"
          >
            <div
              class="go720639114"
            >
              Enter your encryption key:
            </div>
            <div
              class="go3253370986"
            >
              <input
                class="go3854179087"
                name="encryptionKey"
                placeholder="e.g. MyStrongPassword1234"
                type="password"
              />
            </div>
          </section>
        </div>
      </form>
    </div>
    <div
      class="go444860531"
    >
      <button
        class="go3135434547"
        type="button"
      >
        <span
          class="go4062099715"
        >
          Cancel
        </span>
      </button>
      <button
        class="go3135434547"
        disabled=""
        type="button"
      >
        <span
          class="go4062099715"
        >
          Remove
        </span>
      </button>
    </div>
  </section>
</DocumentFragment>
`;

exports[`PasswordEntity when selected when eye icon button has been clicked renders prompt 1`] = `
<DocumentFragment>
  <button
    class="go1778638100"
    isselected="true"
  >
    <span
      class="go1470750214"
    >
      <div
        class="go4131202203"
      >
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Label
          </span>
           - 
          <strong
            class="go1914257566"
          >
            My EA Account
          </strong>
        </div>
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Login
          </span>
           - 
          <strong
            class="go1914257566"
          >
            XXX
          </strong>
        </div>
        <div
          class="go3306043453"
        >
          <span
            class="go3514387725"
          >
            Password
          </span>
           - 
          <strong
            class="go1914257566"
          >
            XXX
          </strong>
        </div>
      </div>
      <div
        class="go1569473757"
      />
    </span>
  </button>
  <button
    class="go3135434547 go3618827259"
  >
    <span
      class="go4062099715"
    >
      <img
        alt="icon"
        class="go115679"
        height="30"
        src="static-asset-stub"
        width="30"
      />
    </span>
  </button>
  <button
    class="go3135434547 go3618827259"
  >
    <span
      class="go4062099715"
    >
      <img
        alt="icon"
        class="go115679"
        height="30"
        src="static-asset-stub"
        width="30"
      />
    </span>
  </button>
  <section
    class="go939053875"
  >
    <div
      class="go579113919"
    >
      <form>
        <div
          class="go4083855640"
        >
          <section
            class="go3154290440"
          >
            <div
              class="go720639114"
            >
              Enter your encryption key:
            </div>
            <div
              class="go3253370986"
            >
              <input
                class="go3854179087"
                name="encryptionKey"
                placeholder="e.g. MyStrongPassword1234"
                type="password"
              />
            </div>
          </section>
        </div>
      </form>
    </div>
    <div
      class="go444860531"
    >
      <button
        class="go3135434547"
        type="button"
      >
        <span
          class="go4062099715"
        >
          Cancel
        </span>
      </button>
      <button
        class="go3135434547"
        disabled=""
        type="button"
      >
        <span
          class="go4062099715"
        >
          Decrypt
        </span>
      </button>
    </div>
  </section>
</DocumentFragment>
`;
